File: Ray Tracing in One Weekend/util/sphere.py
import math
from util.hittable import Hittable
from util.vec3 import dot

class Sphere(Hittable):

    def __init__(self, center, radius):
        self.center = center
        self.radius = max(0.0, radius)

    def hit(self, r, ray_t, rec):

        oc = self.center - r.origin()

        a = dot(r.direction(), r.direction())
        h = dot(r.direction(), oc)
        c = dot(oc, oc) - self.radius * self.radius

        discriminant = h*h - a*c

        if discriminant < 0:
            return False

        sqrtd = math.sqrt(discriminant)

        # nearest root
        root = (h - sqrtd) / a

        if not ray_t.surrounds(root):
            root = (h + sqrtd) / a

            if not ray_t.surrounds(root):
                return False

        rec.t = root
        rec.p = r.at(root)

        outward_normal = (rec.p - self.center) / self.radius
        rec.normal = outward_normal

        return True

File: Ray Tracing in One Weekend - Javascript/Ch8/sphere.js
class Sphere extends Hittable{
   
    constructor(center, radius){
    super()
    this.center = center
    this.radius = Math.max(0,radius)
   }

   hit(r,ray_t,hit_record){
    //super(r,ray_tmin,ray_tmax,rec)
        var oc = Vec3.sub(this.center,r.origin())
        var a = r.direction().length_squared();
        var h = Vec3.dot(r.direction(), oc);
        var c = oc.length_squared() - this.radius*this.radius;
        var discriminant = h*h - a*c;
    
        if (discriminant < 0) {
            return false;
        } 
        
        var sqrtd = Math.sqrt(discriminant);


        var root = (h - sqrtd) / a;
      
        /*
        if (root <= ray_tmin || ray_tmax <= root) {
            root = (h + sqrtd) / a;
            if (root <= ray_tmin || ray_tmax <= root)
                return false;
        }
            */

        if (!ray_t.surrounds(root)) {
            root = (h + sqrtd) / a;
            if (!ray_t.surrounds(root))
                return false;
        }



        var rec = new HitRecord()
        rec.t = root 
        rec.p = r.at(rec.t);
        var outward_normal = Vec3.div(Vec3.sub(rec.p,this.center),this.radius)
        rec.set_face_normal(r,outward_normal)
        hit_record(rec)

        return true
   }

}